### ISSUE TITLE
Alternative approach to handling ipset sets

### ISSUE THREAD

**Issue by NickJH** [DPsystems/web-shield#2]
Can I suggest a couple of different things you can do with loading ipset sets?

1. When you add an entry to a set use the `-exist` flag in case of duplicates.
2. Never load in to the set you want to use. This sounds counter-intuitive, but if you are loading into a set `web-shield`, instead load into a set `web-shield-temp`. Then when it is loaded, swap `web-shield-temp` with `web-shield` and delete `web-shield-temp`

```
# ensure that ipsets exist
ipset destroy -q web-shield-temp
ipset -N web-shield-temp nethash -exist
ipset -N web-shield nethash -exist

# load ipset sets here

# swap sets
ipset swap web-shield web-shield-temp

# ensure that temp sets do not exist
ipset destroy -q "web-shield-temp"
```

You can then safely run your script multiple times without duplications.

3. In your firewall script also create an empty ipset set so it does not fail if you have not created your sets. Then the two scripts can be run in any order.

**Comment by DPsystems** [comment]
Thanks for the feedback.

Does the -exist option not notify you if it's a dupe?  I like that feature so that when I'm adding, I can tell if I have an address duped.

Is it really necessary to use a temp?    Actually, I think I see what you're getting at... if there are entries erased from the blacklist, using a temp when updating would make those disappear - good idea!  I'll work on implementing that!

**Comment by NickJH** [comment]
I don't think you get a message with the `-exist` flag. I use it for scripting in cron so I don't really want a message.

Using a temp allows you to update a set with the firewall rule already running. Otherwise you can't delete the set and have to clear it line by line. There is more I've done to script my blocks to work on boot as well.

**Comment by DPsystems** [comment]
My standard approach is to delete the set and then re-create the set.  I know this probably leaves a little protection window open, but that's not a big deal IMO.  I am also concerned about optimizing memory - it might be more efficient to delete the set and re-create it than hold two sets in RAM?

**Comment by NickJH** [comment]
But you can't delete the set while it is being used in iptables.

re the RAM, once you swap the sets over, you can delete the temp (the old live set). See my code snippet. Extra memory usage is transient. Sets don't take much memory anyway.

**Comment by DPsystems** [comment]
I have revised my START script to act better.  And I created a new RESTART.web-shield.sh script. I am concerned about whether or not creating two parallel sets is less memory efficient than deleting an existing set and recreating it.

I'm open to comments and criticisms, or better approaches.  Let me know what you think?
